### PR TITLE
Added Observable.interval operator

### PR DESCRIPTION
**Pull request by dmitryustimov**

### CODE REVIEW

**Comment by arkivanov** [comment]
Sorry for the delay, I'm sick and will review later. Thanks!

File: reaktive-test/src/commonMain/kotlin/com/badoo/reaktive/test/scheduler/TestScheduler.kt [hunk 1]
```diff
@@ -2,24 +2,26 @@ package com.badoo.reaktive.test.scheduler
...
-    private val _executors: AtomicReference<List<Executor>> = AtomicReference(emptyList(), true)
+    private val _executors: AtomicList<Executor> = atomicList()
     val executors: List<Executor> get() = _executors.value
 
     override fun newExecutor(): Scheduler.Executor =
         Executor(timer, isManualProcessing)
-            .also { executor ->
-                _executors.update { it + executor }
-            }
+            .also { executor -> _executors += executor }
```
Comment (arkivanov) [inline reaktive-test/src/commonMain/kotlin/com/badoo/reaktive/test/scheduler/TestScheduler.kt]: `.also(_executors::add)`

File: reaktive-test/src/commonMain/kotlin/com/badoo/reaktive/test/scheduler/TestScheduler.kt [hunk 2]
```diff
@@ -87,11 +90,26 @@ class TestScheduler(
         fun process() {
             val timeMillis = timer.millis
 
-            tasks
-                .getAndUpdate { it.removeExpiredTasks(timeMillis) }
-                .asSequence()
-                .filter { it.startMillis <= timeMillis }
-                .forEach { it.task() }
+            while (true) {
```
Comment (arkivanov) [inline reaktive-test/src/commonMain/kotlin/com/badoo/reaktive/test/scheduler/TestScheduler.kt]: Please explain why did you change it? Previously it first atomically removed all the expired task and and then executed them all. This is because tasks execution may cause recursion.

File: reaktive/src/commonMain/kotlin/com/badoo/reaktive/observable/Interval.kt [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+import com.badoo.reaktive.utils.atomic.AtomicLong
+
+fun observableInterval(startDelayMillis: Long, periodMillis: Long, scheduler: Scheduler): Observable<Long> =
+    observable { emitter ->
+        val executor = scheduler.newExecutor()
+        emitter.setDisposable(executor)
+
+        val count = AtomicLong()
+        executor.submitRepeating(startDelayMillis, periodMillis) {
+            emitter.onNext(count.value)
```
Comment (arkivanov) [inline reaktive/src/commonMain/kotlin/com/badoo/reaktive/observable/Interval.kt]: Let's initialize AtomicLong with -1L and then just `emitter.onNext(count.addAndGet(1L))`. It's better to use its return value instead of asking separately.

File: reaktive-test/src/commonMain/kotlin/com/badoo/reaktive/test/scheduler/TestScheduler.kt (hunk 2 above)
Comment (dmitryustimov) [reply reaktive-test/src/commonMain/kotlin/com/badoo/reaktive/test/scheduler/TestScheduler.kt]: I've found that it's impossible to check all result of periodic tasks after some time has spent. 
For example, when I observe result of `observableInterval(startDelayMillis = 0L, periodMillis = 100L, scheduler = scheduler)` expected values are `[0, 1, 2, 3]` after 400 ms. But calling `timer.advanceBy(400L)` gives me `[0]` as result of `observableInterval`.
This happens because previous implementation executes tasks that were already scheduled and ignores periodic tasks that can be queued between initial and provided time.
However, result of `observableInterval(startDelayMillis = 0L, periodMillis = 100L, scheduler = scheduler)` will be `[0, 1, 2, 3]` when `timer.advanceBy(100L)` is called 4 times for previous implementation

Comment (dmitryustimov) [reply reaktive-test/src/commonMain/kotlin/com/badoo/reaktive/test/scheduler/TestScheduler.kt]: Of course, my implementation can cause overflow when periodic tasks have 0 ms period.

Comment (arkivanov) [reply reaktive-test/src/commonMain/kotlin/com/badoo/reaktive/test/scheduler/TestScheduler.kt]: How about the following patch to the original implementation?
```
Index: reaktive-test/src/commonMain/kotlin/com/badoo/reaktive/test/scheduler/TestScheduler.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- reaktive-test/src/commonMain/kotlin/com/badoo/reaktive/test/scheduler/TestScheduler.kt	(date 1565972495000)
+++ reaktive-test/src/commonMain/kotlin/com/badoo/reaktive/test/scheduler/TestScheduler.kt	(date 1566041524000)
@@ -91,7 +91,15 @@
                 .getAndUpdate { it.removeExpiredTasks(timeMillis) }
                 .asSequence()
                 .filter { it.startMillis <= timeMillis }
-                .forEach { it.task() }
+                .forEach { task ->
+                    if (task.periodMillis == null) {
+                        task.task()
+                    } else {
+                        repeat(((timeMillis - task.startMillis) / task.periodMillis).toInt() + 1) {
+                            task.task()
+                        }
+                    }
+                }
         }
 
         private fun processIfNeeded() {
@@ -121,7 +129,7 @@
                 mapNotNull {
                     when {
                         it.startMillis > timeMillis -> it
-                        it.periodMillis != null -> it.copy(startMillis = timeMillis + it.periodMillis)
+                        it.periodMillis != null -> it.copy(startMillis = (timeMillis - it.startMillis) % it.periodMillis)
                         else -> null
                     }
                 }

```

Comment (arkivanov) [reply reaktive-test/src/commonMain/kotlin/com/badoo/reaktive/test/scheduler/TestScheduler.kt]: I covered it locally with unit tests, will create a PR after this will be merged

Comment (dmitryustimov) [reply reaktive-test/src/commonMain/kotlin/com/badoo/reaktive/test/scheduler/TestScheduler.kt]: Sorry for the delay. It seems that suggested changes does not keep an order of execution and ignores time span provided through `timer` less than task's period.
You can look up at IntervalTest#emit_values_periodically_AFTER_task_executed_few_times - it  fails with the suggested patch.

Comment (arkivanov) [reply reaktive-test/src/commonMain/kotlin/com/badoo/reaktive/test/scheduler/TestScheduler.kt]: Yup, you are right. Sorry for that. Here is the updated patch. Please apply to the master's TestScheduler:
```
Index: reaktive-test/src/commonMain/kotlin/com/badoo/reaktive/test/scheduler/TestScheduler.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- reaktive-test/src/commonMain/kotlin/com/badoo/reaktive/test/scheduler/TestScheduler.kt	(date 1565972495000)
+++ reaktive-test/src/commonMain/kotlin/com/badoo/reaktive/test/scheduler/TestScheduler.kt	(date 1566255120000)
@@ -2,9 +2,14 @@
 
 import com.badoo.reaktive.scheduler.Scheduler
 import com.badoo.reaktive.utils.atomic.AtomicBoolean
+import com.badoo.reaktive.utils.atomic.AtomicList
 import com.badoo.reaktive.utils.atomic.AtomicLong
 import com.badoo.reaktive.utils.atomic.AtomicReference
+import com.badoo.reaktive.utils.atomic.add
+import com.badoo.reaktive.utils.atomic.atomicList
+import com.badoo.reaktive.utils.atomic.clear
 import com.badoo.reaktive.utils.atomic.getAndUpdate
+import com.badoo.reaktive.utils.atomic.plusAssign
 import com.badoo.reaktive.utils.atomic.update
 
 class TestScheduler(
@@ -12,14 +17,12 @@
 ) : Scheduler {
 
     val timer = Timer()
-    private val _executors: AtomicReference<List<Executor>> = AtomicReference(emptyList(), true)
+    private val _executors: AtomicList<Executor> = atomicList()
     val executors: List<Executor> get() = _executors.value
 
     override fun newExecutor(): Scheduler.Executor =
         Executor(timer, isManualProcessing)
-            .also { executor ->
-                _executors.update { it + executor }
-            }
+            .also(_executors::add)
 
     override fun destroy() {
         _executors
@@ -57,7 +60,7 @@
         private val isManualProcessing: Boolean
     ) : Scheduler.Executor {
 
-        private val tasks: AtomicReference<List<Task>> = AtomicReference(emptyList(), true)
+        private val tasks: AtomicList<Task> = atomicList()
         private val _isDisposed = AtomicBoolean()
         override val isDisposed: Boolean get() = _isDisposed.value
         private val timerListener = ::processIfNeeded
@@ -75,7 +78,7 @@
         }
 
         override fun cancel() {
-            tasks.value = emptyList()
+            tasks.clear()
         }
 
         override fun dispose() {
@@ -86,12 +89,21 @@
 
         fun process() {
             val timeMillis = timer.millis
+            val tasksToStart = ArrayList<() -> Unit>()
+            val queue = tasks.value.toMutableList()
+
+            while (queue.isNotEmpty() && (queue.first().startMillis <= timeMillis)) {
+                val task = queue.removeAt(0)
+                tasksToStart += task.task
+                if (task.periodMillis != null) {
+                    queue += task.copy(startMillis = task.startMillis + task.periodMillis)
+                    queue.sort()
+                }
+            }
 
-            tasks
-                .getAndUpdate { it.removeExpiredTasks(timeMillis) }
-                .asSequence()
-                .filter { it.startMillis <= timeMillis }
-                .forEach { it.task() }
+            tasks.value = queue
+
+            tasksToStart.forEach { it() }
         }
 
         private fun processIfNeeded() {
@@ -115,17 +127,6 @@
 
             processIfNeeded()
         }
-
-        private companion object {
-            private fun List<Task>.removeExpiredTasks(timeMillis: Long): List<Task> =
-                mapNotNull {
-                    when {
-                        it.startMillis > timeMillis -> it
-                        it.periodMillis != null -> it.copy(startMillis = timeMillis + it.periodMillis)
-                        else -> null
-                    }
-                }
-        }
     }
 
     private data class Task(
```
